### ISSUE TITLE
DriverError { Packet out of sync } when reading large payloads with compression enabled

### ISSUE THREAD

**Issue by isegal** [blackbeam/rust-mysql-simple#175]
Steps:
1. Import the following test file to mysql server into a 'test' schema:

https://gist.githubusercontent.com/isegal/0b7cff24253fc64b834ed5d8dc2dc2fe/raw/e0c3eedc4b14798b1be3a133d23aa84c0f8b339d/gistfile1.txt

2. Run the following code that reads the table row by row:

```
#[macro_use]
extern crate mysql;

use mysql as my;
use mysql::{Pool, PooledConn};


fn main() {
    let dest_pool = my::Pool::new_manual(1, 1, "mysql://root:@127.0.0.1:3306/test?compress=true").unwrap();
    let mut conn = dest_pool.get_conn().unwrap();
    read_test_table(&mut conn);
}

fn read_test_table(conn: &mut PooledConn) {
    let result = conn.prep_exec("SELECT * FROM field_test ORDER BY id ASC", ()).unwrap();

    for row in result {
        let values = row.unwrap().unwrap();
        dbg!(&values[0]);
    }
}
```

Suspected culprit:

According to the protocol specification, the compressed packet sequence id should be incremented independently from the regular packet id.

https://dev.mysql.com/doc/internals/en/compressed-packet-header.html

Therefore, the compressed sequence id number should not cross over from compressed sequence id to regular sequence id:

https://github.com/blackbeam/rust-mysql-simple/blob/master/src/io/mod.rs#L486

https://github.com/blackbeam/rust-mysql-simple/blob/master/src/conn/mod.rs#L862

In the run below, I added dbg() to print sequence IDs.
As the server sends comp_sequence_id up to 14, it then gets assigned to old_sequence_id which then gets passed to regular read packet and fails the sanity check:

```

....
[rust-mysql-simple/src/conn/mod.rs:863] "Compressed" = "Compressed"
[rust-mysql-simple/src/io/mod.rs:284] (&srv_seq_id, &seq_id) = (
    92,
    92
)
[rust-mysql-simple/src/conn/mod.rs:871] &seq_id = 93
[src/main.rs:19] &values[0] = Int(
    1013652
)
[rust-mysql-simple/src/conn/mod.rs:863] "Compressed" = "Compressed"
[rust-mysql-simple/src/io/mod.rs:284] (&srv_seq_id, &seq_id) = (
    93,
    93
)
[rust-mysql-simple/src/conn/mod.rs:871] &seq_id = 94
[src/main.rs:19] &values[0] = Int(
    1013653
)
[rust-mysql-simple/src/conn/mod.rs:863] "Compressed" = "Compressed"
[rust-mysql-simple/src/io/mod.rs:409] comp_seq_id = 14
[rust-mysql-simple/src/io/mod.rs:284] (&srv_seq_id, &seq_id) = (
    94,
    14
)
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: DriverError { Packet out of sync }', src/libcore/result.rs:997:5
```

**Comment by blackbeam** [comment]
Hi. Thanks for great report!

> Therefore, the compressed sequence id number should not cross over from compressed sequence id to regular sequence id:

I've found this part of protocol specification confusing because in practice this is not true ([see here][1]). This features was quite tricky to implement because of this (see 883b91f and e69cc9a).

I hope it's finally fixed in 3f8cb87. I'll release it soon.

[1]: https://github.com/mysql/mysql-server/tree/ab8ebdf41781ba31e3ad755d05c2063d15176100/sql-common#L226

**Comment by isegal** [comment]
I just ran a test on a large table, it looks like the fix works, thank you.

**Comment by blackbeam** [comment]
Fix published in [v16.0.2](https://github.com/blackbeam/rust-mysql-simple/releases/tag/v16.0.2) and [v15.1.2](https://github.com/blackbeam/rust-mysql-simple/releases/tag/v15.1.2)
